text/3857-cfg-version.md

```diff
@@ -0,0 +1,625 @@
+
+Accessible
+- Channel
+- Commit date
+- Commit hash
+- LLVM version
+- Version
+
+## Other
+
+Python
+- Programmatic version: [`sys.version_info`](https://docs.python.org/3/library/sys.html#sys.version_info)
+- Vendor display version: [`sys.version`](https://docs.python.org/3/library/sys.html#sys.version)
+- [Dependency specifiers](https://packaging.python.org/en/latest/specifications/dependency-specifiers/)
+  - e.g. `requests [security,tests] >= 2.8.1, == 2.8.* ; python_version < "2.7"`
+
+C++
+- Numeric value representing the version of the C++ standard: [`__cplusplus`](https://en.cppreference.com/w/cpp/preprocessor/replace):
+
+C:
+- Implementation-defined value of the C++ standard: [__STDC_VERSION__](https://en.cppreference.com/w/cpp/preprocessor/replace)
+
+Haskell:
+- Numeric value representing the vendor's version, e.g. `#if __GLASGOW_HASKELL__ >= 706`
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- In the future the `--check-cfg` `since()` predicate could make the minimum-version field optional,
+  matching all version numbers.
+
+## Relaxing SemVer
+
+Instead of requiring the `IDENTIFIER` in the `check-cfg` `since` predicate to be strictly SemVer `major.minor.patch`,
+we could allow abbreviated forms like `major.minor` or even `major`.
+This would make the predicate more inclusive for other cases, like `edition`.
+
+## Vendor name and version
+
+We could add `--cfg`s for the compiler vendor name and version.
+In addition to the use cases given in the Motivation section,
+this will allow users to check for specific nightly versions.
+
+Some challenges for this with `rustc --version`:
+- Nightly versions for a given release are mutable,
+  all mapping to the `-nightly` pre-release version rather than including the date within the pre-release
+- This does not conform to SemVer's precedence rules,
+  as `-nightly` is an older version than `-beta.4` while [SemVer's precedence rules](https://semver.org/#spec-item-11) say the opposite
+- Crater runs and local builds don't necessarily have a version that fits within this picture
+
+## `#[cfg(nightly)]`
+
+Depending on what is meant by this,
+we either need the language version or the vendor name and version as well as a way to check for the presence of `pre-release`.
+
+See also [`#[cfg(nightly)]`](https://rust-lang.github.io/rfcs/2523-cfg-path-version.html#cfgnightly) in the previous RFC.
+
+## `cfg_target_version`
+
+Instead of defining a new `#[cfg]` predicate, [RFC 3750](https://github.com/rust-lang/rfcs/pull/3750)
+could reuse the `#[cfg(since)]` predicate.
+
+As not all systems use SemVer, we can either
+- Contort the version into SemVer
+  - This can run into problems either with having more precision (e.g. `120.0.1.10` while SemVer only allows `X.Y.Z`) or post-release versions (e.g. `1.2.0.post1` which a SemVer predicate would treat as a pre-release).
```

Comment:
This is seemingly self-conflicting: `1.2.0.post1` is actually just invalid SemVer per the first half of the sentence, as SemVer requires prerelease versions to be separated by a hyphen. Perhaps you meant
```suggestion
  - This can run into problems either with having more precision (e.g. `120.0.1.10` while SemVer only allows `X.Y.Z`) or post-release versions (e.g. `1.2.0-post1` which a SemVer predicate would treat as a pre-release).
```

Reply:
It might be worth noting that "120.0.1-10" would work, and would compare correctly against "120.0.1-11", "120.0.1-9", and such, due to the details of SemVer's prerelease ordering allowing multiple dot-separated identifiers in a prerelease version, and ordering by integer sort if a given identifier is entirely digits.

But you don't need to copy-paste the SemVer docs into this RFC if that information seems extraneous.

Reply:
I tried clarifying this in 4352a955.